x/gamm/keeper/migrate.go

```diff
@@ -0,0 +1,142 @@
+	osmoutils.MustSet(store, types.KeyMigrationInfo, &migrationInfo)
+}
+
+// validateRecords validates a list of BalancerToConcentratedPoolLink records to ensure that:
+// 1) there are no duplicates,
+// 2) the records are in sorted order.
+// 3) both the balancer and gamm pool IDs are valid
+func (k Keeper) validateRecords(ctx sdk.Context, records ...types.BalancerToConcentratedPoolLink) error {
+	lastBalancerPoolID := uint64(0)
+	balancerIdFlags := make(map[uint64]bool)
+	clIdFlags := make(map[uint64]bool)
+
+	for _, record := range records {
+		// If the balancer ID has already been seen, we have a duplicate
+		if balancerIdFlags[record.BalancerPoolId] {
+			return fmt.Errorf(
+				"Balancer pool ID #%d has duplications.",
+				record.BalancerPoolId,
+			)
+		}
+
+		// If the concentrated ID has already been seen, we have a duplicate
+		if clIdFlags[record.ClPoolId] {
+			return fmt.Errorf(
+				"Concentrated pool ID #%d has duplications.",
+				record.ClPoolId,
+			)
+		}
+
+		// Ensure records are sorted from lowest to highest balancer pool ID
+		if record.BalancerPoolId < lastBalancerPoolID {
+			return fmt.Errorf(
+				"Balancer pool ID #%d came after Balancer pool ID #%d.",
+				record.BalancerPoolId, lastBalancerPoolID,
+			)
+		}
+
+		// Ensure the provided balancerPoolId exists and that it is of type balancer
+		poolType, err := k.GetPoolType(ctx, record.BalancerPoolId)
+		if err != nil {
+			return err
+		}
+		if poolType.String() != "Balancer" {
+			return fmt.Errorf("Balancer pool ID #%d is not of type balancer", record.BalancerPoolId)
+		}
+
+		// Ensure the concentrated pool exists. If clPoolID is 0, this signals a removal, so we skip this check.
+		// TODO: Get GetPoolType to work for cl pools from gamm
+		if record.ClPoolId != 0 {
+			_, err = k.poolManager.GetPoolModule(ctx, record.ClPoolId)
+			if err != nil {
+				return err
+			}
+		}
+
+		lastBalancerPoolID = record.BalancerPoolId
+
+		balancerIdFlags[record.BalancerPoolId] = true
+		clIdFlags[record.ClPoolId] = true
+	}
+	return nil
+}
+
+// ReplaceMigrationRecords gets the current migration records and replaces it in its entirety with the provided records.
+// It is checked for no err when a proposal is made, and executed when a proposal passes.
+func (k Keeper) ReplaceMigrationRecords(ctx sdk.Context, records ...types.BalancerToConcentratedPoolLink) error {
+	migrationInfo := k.GetMigrationInfo(ctx)
+
+	err := k.validateRecords(ctx, records...)
+	if err != nil {
+		return err
+	}
+
+	migrationInfo.BalancerToConcentratedPoolLinks = records
+
+	k.SetMigrationInfo(ctx, migrationInfo)
+	return nil
+}
+
+// UpdateDistrRecords gets the current migration records and only updates the records that are provided.
+// It is checked for no err when a proposal is made, and executed when a proposal passes.
+func (k Keeper) UpdateMigrationRecords(ctx sdk.Context, records ...types.BalancerToConcentratedPoolLink) error {
+	recordsMap := make(map[uint64]types.BalancerToConcentratedPoolLink)
```

Comment:
nit: can be pre-allocated

Reply:
https://github.com/osmosis-labs/osmosis/pull/4071/commits/078e86f98ca0813e025b58f588fe7f4e43c9e83e